### PR TITLE
Feature/refactor

### PR DESCRIPTION
**Pull request by maciejburda**
Multiple improvements related to Friday architecture meeting:
- expiring pub/sub
- delivered matches are a union of sets
- deleting device will delete related pubs/subs
- delete all is now using `DispatchGroup` to deliver notification when all deletions are finished
- other on the fly improvements

### CODE REVIEW

File: Example/Alps/MatchDelegate.swift [hunk 1]
```diff
@@ -8,10 +8,11 @@
 
 import Foundation
 import AlpsSDK
+import Alps
 
 class MatchDelegate: AlpsManagerDelegate {
-    var onMatch: OnMatchClojure
-    init(_ onMatch: @escaping OnMatchClojure) {
```
Comment (tharpa) [inline Example/Alps/MatchDelegate.swift:14]: You deleted my little joke lol

**Comment by tharpa** [comment]
Hey @maciejburda, well I've just skimmed through this; I trust @wenjdu will review it more in detail.

File: AlpsSDK/MatchMonitor.swift [hunk 2]
```diff
@@ -54,12 +54,15 @@ public class MatchMonitor: RemoteNotificationManagerDelegate {
         guard let deviceId = device.id else { return }
         MatchesAPI.getMatches(deviceId: deviceId) { (matches, error) in
             guard let matches = matches, matches.count > 0, error == nil else { return }
-            self.deliveredMatches = Set(matches)
-            self.delegate?.didFind(matches: matches, for: device)
+            let union = self.deliveredMatches.union(Set(matches))
+            if union != self.deliveredMatches {
+                self.deliveredMatches = union
+                self.delegate?.didFind(matches: matches, for: device)
```
Comment (wenjdu) [inline AlpsSDK/MatchMonitor.swift:60]: great, so now if matches are already delivered they won't be a second time right ?

File: AlpsSDK/Repositories/PubSub/PublicationRepository.swift [hunk 3]
```diff
@@ -11,17 +11,27 @@ import Alps
...
+        }
+        set {
+            _items = newValue
+        }
+    }
+    
+    private var _items = [Publication]() {
         didSet {
-            _ = PersistenceManager.save(object: self.items.map { $0.encodablePublication }, to: kPublicationFile)
+            _ = PersistenceManager.save(object: self._items.withoutExpired.map { $0.encodablePublication }, to: kPublicationFile)
```
Comment (wenjdu) [inline AlpsSDK/Repositories/PubSub/PublicationRepository.swift:29]: It is good that we also keep the item that are expired, there might be some interesting use-case. 
I like your implementation 👍

File: Example/Alps/AppDelegate.swift [hunk 4]
```diff
@@ -14,26 +14,53 @@ import AlpsSDK
...
+            // Create New Subscription
+            let subscription = Subscription(topic: "Test Topic", range: 20, duration: 100, selector: "test = 'true'")
+            self.alpsManager.createSubscription(subscription: subscription, completion: { _ in
+                print("Subscription was created \(subscription)")
+            })
+            
+            // Start Monitoring
+            self.matchDelegate = MatchDelegate { matches, _ in
+                print("You've got a match!\n\(matches)")
+                UIApplication.shared.registerForRemoteNotifications()
```
Comment (wenjdu) [inline Example/Alps/AppDelegate.swift:45]: Why do you register for Remote Notifications here ??

**Review by wenjdu — COMMENTED** [review]
Good job, I will have a try with Ticketing App when you merge 🥇

File: Example/Alps/MatchDelegate.swift (hunk 1 above)
Comment (maciejburda) [reply Example/Alps/MatchDelegate.swift:14]: I remember when reading this line I was like: "When did that happen?" :)

File: Example/Alps/AppDelegate.swift (hunk 4 above)
Comment (maciejburda) [reply Example/Alps/AppDelegate.swift:45]: Good catch. It's something that I forgot to remove when testing APNS with Łukasz.

If you don't mind - I will get back to this example in next PR.

Comment (wenjdu) [reply Example/Alps/AppDelegate.swift:45]: No problem ! I saw that you comment the registerForPushNotifications() in `RemoteNotificationManager` class. 
Did you succeed in getting the device token ?

**Pull request merged**